sphinx/ext/intersphinx/_load.py

```diff
@@ -136,36 +137,44 @@ def load_mappings(app: Sphinx) -> None:
     intersphinx_cache: dict[InventoryURI, InventoryCacheEntry] = inventories.cache
     intersphinx_mapping: IntersphinxMapping = app.config.intersphinx_mapping
 
+    expected_uris = {uri for uri, _invs in app.config.intersphinx_mapping.values()}
+
+    # If the current cache contains some (project, uri) pair
+    # say ("foo", "foo.com") and if the new intersphinx dict
+    # contains the pair ("foo", "bar.com"), we need to remove
+    # the ("foo", "foo.com") entry and use ("foo", "bar.com").
+    for uri in frozenset(intersphinx_cache):
+        if intersphinx_cache[uri][0] not in intersphinx_mapping or uri not in expected_uris:
+            # remove a cached inventory if the latter is no more used by intersphinx
+            del intersphinx_cache[uri]
+
     with concurrent.futures.ThreadPoolExecutor() as pool:
-        futures = []
-        for name, (uri, invs) in intersphinx_mapping.values():
-            futures.append(pool.submit(
-                fetch_inventory_group, name, uri, invs, intersphinx_cache, app, now,
-            ))
+        futures = [
+            pool.submit(fetch_inventory_group, name, uri, invs, intersphinx_cache, app, now)
+            for name, (uri, invs) in app.config.intersphinx_mapping.values()
+        ]
         updated = [f.result() for f in concurrent.futures.as_completed(futures)]
 
     if any(updated):
+        # clear the local inventories
         inventories.clear()
 
         # Duplicate values in different inventories will shadow each
-        # other; which one will override which can vary between builds
-        # since they are specified using an unordered dict.  To make
-        # it more consistent, we sort the named inventories and then
-        # add the unnamed inventories last.  This means that the
-        # unnamed inventories will shadow the named ones but the named
-        # ones can still be accessed when the name is specified.
-        named_vals = []
-        unnamed_vals = []
-        for name, _expiry, invdata in intersphinx_cache.values():
-            if name:
-                named_vals.append((name, invdata))
-            else:
-                unnamed_vals.append((name, invdata))
-        for name, invdata in sorted(named_vals) + unnamed_vals:
-            if name:
-                inventories.named_inventory[name] = invdata
-            for type, objects in invdata.items():
-                inventories.main_inventory.setdefault(type, {}).update(objects)
+        # other and which one will override which varies between builds.
+        #
+        # We can however order the cache by URIs for reproducibility.
+        intersphinx_cache_values = sorted(intersphinx_cache.values(), key=itemgetter(0, 1))
```

Comment:
Making it a bit clearer what the sort keys are here could be useful - at the moment I think there's eventually an `Inventory` (dict) object used as a comparison tie-breaker?

Reply:
Unfortunately, I don't know how you could do it since dicts are not comparable =/

Reply:
It's my mistake - I thought that the `(0, 1)` lookups evaluated to `InventoryURI, InventoryCacheEntry` (looking at the typehint without registering the effect of `.values()`).  That would mean that the cache entry in its' entirety would be used as a comparator -- and indeed that would fail as you mention.